docs/cunumeric/source/user/practices.rst

```diff
@@ -0,0 +1,411 @@
+implementation instead of an explicit loop-based implementation.
+
+.. code-block:: python
+
+    # x and y are three-dimensional arrays
+
+    # Not recommended: Naive element-wise implementation
+    for i in range(ny):
+        for j in range(nx):
+            x[0, j, i] = y[3, j, i]
+
+    # Recommended: Array-based implementation
+    x[0] = y[3]
+
+The same recommendation applies when the value we are setting to is a scalar
+or when values are set conditionally. We first form the condition array
+corresponding to the conditional and then use that to update the array,
+essentially breaking it down to three steps:
+
+* create the condition array
+* update the array corresponding to the `if` statement
+* update the array corresponding to the `else` statement while noting that
+  the condition is flipped for the `else` statement.
+
+.. code-block:: python
+
+    # x and y are two-dimensional arrays, and we need to update x
+    # depending on whether y meets a condition or not.
+
+    # Not recommended: Naive element-wise implementation
+    for i in range(ny):
+        for j in range(nx):
+            if (y[j, i] < tol):
+                x[j, i] = const
+            else
+                x[j, i] = 1.0 - const
+
+    # Recommended: Array-based implementation
+    cond = y < tol
+    x[cond] = const
+    x[~cond] = 1.0 - const
+
+In the example below, using a boolean mask array will be faster than using
+indices. For the curious reader, using indices with cuNumeric will require
+additional communication that might be undesirable for performance.
+
+.. code-block:: python
+
+    import cunumeric as np
+
+    # Not recommended: don't use nonzero to get indices
+    indices = np.nonzero(h < 0)
+    x[indices] = y[indices]
+
+    # Recommended: Use boolean mask to update the array
+    cond = h < 0
+    x[cond] = y[cond]
+
+
+When the array needs to be updated from another array based on a condition
+that they both satisfy, use ``putmask`` for better performance. Unlike the
+previous example, here x is set to twice the value of y when the condition
+is met.
+
+.. code-block:: python
+
+    import cunumeric as np
+
+    # We need to update elements of x from y based on a condition
+    cond = y < tol
+
+    # Acceptable
+    x[cond] = y[cond] * 2.0
+
+    # Recommended: use putmask to update elements based on a condition
+    np.putmask(x, cond, y * 2.0)
+
+Logic Functions
+~~~~~~~~~~~~~~~
+
+Setting elements of an array that satisfy multiple conditions to a scalar
+should be done using logic functions instead of iterating through a loop.
+Here is an example:
+
+.. code-block:: python
+
+    # Not recommended: naive element-wise update to update x
+    for i in range(ny):
+        for j in range(nx):
+            if (first_cond and second_cond):
+                x[j, i] = const
+
+    # Recommended: Use logical operations.
+    x[np.logical_and(first_cond, second_cond)] = const
+
+
+Refer to the documentation for other logical operations.
```

Comment:
should we link documentation to this?
https://numpy.org/doc/stable/reference/routines.logic.html#logical-operations